cmd/encryption-v1.go

```diff
@@ -1157,3 +1159,119 @@ func DecryptObjectInfo(info *ObjectInfo, headers http.Header) (encrypted bool, e
 	}
 	return
 }
+
+// The customer key in the header is used by the gateway for encryption in the case of
+// s3 gateway double encryption. A new client key is derived from the customer provided
+// key to be sent to the s3 backend for encryption at the backend.
+func deriveClientKey(clientKey [32]byte, bucket, object string) ([32]byte, error) {
+	var key [32]byte
+	mac := hmac.New(sha256.New, clientKey[:])
+	mac.Write([]byte(crypto.SSEC.String()))
+	mac.Write([]byte(path.Join(bucket, object)))
+	mac.Sum(key[:0])
+	return key, nil
+}
+
+// extract encryption options for pass through to backend in the case of gateway
+func extractEncryptionOption(header http.Header, copySource bool, metadata map[string]string) (opts ObjectOptions, err error) {
+	var clientKey []byte
+	var sse encrypt.ServerSide
+
+	if copySource {
+		if crypto.SSECopy.IsRequested(header) {
+			clientKey, err = base64.StdEncoding.DecodeString(header.Get(crypto.SSECopyKey))
+			if err != nil || len(clientKey) != 32 { // The client key must be 256 bits long
+				return opts, crypto.ErrInvalidCustomerKey
+			}
+			if sse, err = encrypt.NewSSEC(clientKey); err != nil {
+				return
+			}
+			return ObjectOptions{ServerSideEncryption: encrypt.SSECopy(sse)}, nil
+		}
+		return
+	}
+
+	if crypto.SSEC.IsRequested(header) {
+		clientKey, err = base64.StdEncoding.DecodeString(header.Get(crypto.SSECKey))
```

Comment:
Same, already implemented in [`SSEC.ParseHTTP`](https://github.com/minio/minio/blob/master/cmd/crypto/header.go#L174)

Reply:
done